src/rust/engine/workunit_store/src/lib.rs

```diff
@@ -26,17 +26,72 @@
 // Arc<Mutex> can be more clear than needing to grok Orderings:
 #![allow(clippy::mutex_atomic)]
 
+use futures::task_local;
 use parking_lot::Mutex;
+use rand::thread_rng;
+use rand::Rng;
+use std::collections::HashSet;
+use std::sync::Arc;
+use time::Timespec;
 
+#[derive (Clone, Debug, PartialEq, Eq, Hash)]
 pub struct WorkUnit {
   pub name: String,
-  pub start_timestamp: f64,
-  pub end_timestamp: f64,
+  pub start_timestamp: Timespec,
+  pub end_timestamp: Timespec,
   pub span_id: String,
+  pub parent_id: Option<String>,
 }
 
-pub trait WorkUnitStore {
-  fn should_record_zipkin_spans(&self) -> bool;
-  fn get_workunits(&self) -> &Mutex<Vec<WorkUnit>>;
-  fn add_workunit(&self, workunit: WorkUnit);
+#[derive (Clone)]
+pub struct WorkUnitStore {
+  workunits: Arc<Mutex<HashSet<WorkUnit>>>,
+}
+
+impl WorkUnitStore {
+  pub fn new() -> WorkUnitStore {
+    WorkUnitStore {
+      workunits: Arc::new(Mutex::new(HashSet::new())),
+    }
+  }
+
+  pub fn get_workunits(&self) -> Arc<Mutex<HashSet<WorkUnit>>> {
+    self.workunits.clone()
+  }
+
+  pub fn add_workunit(&self, workunit: WorkUnit) {
+    self.workunits.lock().insert(workunit);
+  }
+}
+
+pub fn generate_random_64bit_string() -> String {
+    let mut rng = thread_rng();
+    let random_u64: u64 = rng.gen();
+    format!("{:16.x}", random_u64)
+}
+
+pub fn got_workunits(workunit_store: WorkUnitStore) -> HashSet<WorkUnit> {
```

Comment:
Let's rename this to something like: `workunits_without_span_id` or `workunits_with_constant_span_id`

Also, this only needs a reference to a `WorkUnitStore`, so let's take a `&WorkUnitStore` instead of a `WorkUnitStore`.

There's also a cute little construction where you _could_ if you wanted (I don't think it's necessarily better here, but it's pretty cute and good to know about), write:
```rust
workunit_store.get_workunits().lock().iter()
  .map(|workunit| WorkUnit { span_id: String::from("ignore"), **workunit })
  .collect()
```

which means "Make me a new `WorkUnit` with all the fields set the same as `workunit` except `span_id`"

Reply:
When I changed the code to 
```
 workunit_store.get_workunits().lock().iter().map(|workunit| WorkUnit { span_id: String::from("ignore"), ..workunit }).collect()
```
I got the next error:
```
error[E0308]: mismatched types
  --> workunit_store/src/lib.rs:76:109
   |
76 |   workunit_store.get_workunits().lock().iter().map(|workunit| WorkUnit { span_id: String::from("ignore"), ..workunit }).collect()
   |                                                                                                             ^^^^^^^^ expected struct `WorkUnit`, found &WorkUnit
   |
```

Reply:
It looks like `WorkUnit { span_id: String::from("ignore"), ..workunit }` moves fields from one struct to another
`WorkUnit { span_id: String::from("ignore"), ..workunit.clone() }` works